middleware/hooks.go

```diff
@@ -12,3 +15,133 @@ type Hook interface {
 	HandleScrape(context.Context, *bittorrent.ScrapeRequest, *bittorrent.ScrapeResponse) (context.Context, error)
 }
+
+type skipSwarmInteraction struct{}
+
+// SkipSwarmInteractionKey is a key for the context of an Announce to control
+// whether the swarm interaction middleware should run.
+// Any non-nil value set for this key will cause the swarm interaction
+// middleware to skip.
+var SkipSwarmInteractionKey = skipSwarmInteraction{}
+
+type swarmInteractionHook struct {
+	store storage.PeerStore
+}
+
+func (h *swarmInteractionHook) HandleAnnounce(ctx context.Context, req *bittorrent.AnnounceRequest, resp *bittorrent.AnnounceResponse) (_ context.Context, err error) {
+	if ctx.Value(SkipSwarmInteractionKey) != nil {
+		return ctx, nil
+	}
+
+	switch {
+	case req.Event == bittorrent.Stopped:
+		err = h.store.DeleteSeeder(req.InfoHash, req.Peer)
+		if err != nil && err != storage.ErrResourceDoesNotExist {
+			return ctx, err
+		}
+
+		err = h.store.DeleteLeecher(req.InfoHash, req.Peer)
+		if err != nil && err != storage.ErrResourceDoesNotExist {
+			return ctx, err
+		}
+	case req.Event == bittorrent.Completed:
+		err = h.store.GraduateLeecher(req.InfoHash, req.Peer)
+		return ctx, err
+	case req.Left == 0:
+		// Completed events will also have Left == 0, but by making this
+		// an extra case we can treat "old" seeders differently from
+		// graduating leechers. (Calling PutSeeder is probably faster
+		// than calling GraduateLeecher.)
+		err = h.store.PutSeeder(req.InfoHash, req.Peer)
+		return ctx, err
+	default:
+		err = h.store.PutLeecher(req.InfoHash, req.Peer)
+		return ctx, err
+	}
+
+	return ctx, nil
+}
+
+func (h *swarmInteractionHook) HandleScrape(ctx context.Context, _ *bittorrent.ScrapeRequest, _ *bittorrent.ScrapeResponse) (context.Context, error) {
+	// Scrapes have no effect on the swarm.
+	return ctx, nil
+}
+
+// ErrInvalidIP indicates an invalid IP for an Announce.
+var ErrInvalidIP = errors.New("invalid IP")
+
+// ErrMissingScrapeAddressLength indicates that the ScrapeAddressLengthKey
+// in the context for a Scrape was not set or not an int.
+var ErrMissingScrapeAddressLength = errors.New("missing address length for scrape")
+
+// ErrInvalidScrapeAddressLength indicates that the ScrapeAddressLength given
+// in the context for a scrape was invalid.
+var ErrInvalidScrapeAddressLength = errors.New("invalid scrape address length")
+
+type skipResponseHook struct{}
+
+// SkipResponseHookKey is a key for the context of an Announce or Scrape to
+// control whether the response middleware should run.
+// Any non-nil value set for this key will cause the response middleware to
+// skip.
+var SkipResponseHookKey = skipResponseHook{}
+
+type scrapeAddressLength struct{}
+
+// ScrapeAddressLengthKey is the key under which to store the length of the
+// requester's IP address for a scrape.
+// The value is expected to be of type int.
+var ScrapeAddressLengthKey = scrapeAddressLength{}
```

Comment:
we do actually care about this length or do we care that it's ipv4 or ipv6? maybe this should be an bool instead of an int